### DISCUSSION TITLE
Python 3.8

### DISCUSSION THREAD

**Discussion by neurodavidus** [wehr-lab/autopilot#72] [Q&A] [answered]
Hi there, 
Very basic question. I was planning to install autopilot on a fresh Ubunto 20.04 install. It comes with Python 3.8. 
The installation guideline says python 3.7. Should I downgrade my Python version or should I give it a try (my feeling)
thanks for the help
David

**Comment by sneakers-the-rat** [comment, marked as answer]
The 3.7 dependency is almost entirely due to a proprietary camera driver only being available for 3.7, and i desparately want to dump that dependency. (came up before in the previous discussion board https://groups.google.com/g/autopilot-users/c/o--7rNH9j9w/m/PIG9G0JgBQAJ )

If you do try to venture on to installing 3.7, I would recommend using pyenv https://github.com/pyenv/pyenv to manage python installations. Otherwise, you could comment out this line to allow you to use 3.8 (it should work, but I haven't been testing using 3.8!) : https://github.com/wehr-lab/autopilot/blob/e82dfc27ba3bafad72c3a5bd5d3067442a072dfc/setup.py#L149

**Reply by neurodavidus** [reply]
OK, thanks. I installed 3.7 through pyenv (in fact I had to choose a specific version so I chose 3.7.4) because, in one of my first projects, I want to acquire videos using a PointGrey Camera, which I think requires Spinnaker.
I Installed autopilot and run the configuration for the terminal but I stop there because I was not sure of what I am doing ...next step is the pilot.

**Reply by sneakers-the-rat** [reply]
Fabulous! I've only used FLIR cameras with it, but theoretically should work with any GenICam compliant camera. For installing spinnaker, see this script: https://github.com/wehr-lab/autopilot/blob/main/autopilot/setup/install_pyspin.sh , and then the relevant hardware class would be this: https://docs.auto-pi-lot.com/en/latest/autopilot.hardware.cameras.html#autopilot.hardware.cameras.Camera_Spinnaker and the various methods that it overrides are described more fully in the base camera class: https://docs.auto-pi-lot.com/en/latest/autopilot.hardware.cameras.html#autopilot.hardware.cameras.Camera

We'll probably have to do some tweaking to get your camera to work, but happy to do that with you! Would love to make that class do what you want it to do :)

**Reply by neurodavidus** [reply]
Hi Guys
A bit confused here. I wanted to reinstall autopilot from the source on the terminal computer (originally I installed using pip3)
I pip3 uninstall fine: then I did the following
![image](https://user-images.githubusercontent.com/4154936/108094256-e845a280-707e-11eb-828c-29018aa4285b.png)

So I am a bit confused. Also if I reinstall using pip3, autopilot installs just fine with or without setting pyenv on 3.7.4. 
thanks for the help
DAvid

**Reply by sneakers-the-rat** [reply]
seems like a few things happening here with `$PATH`, I think best way to describe might be with a bit more information on pyenv and virtualenvs. So you can think of the system python binaries, pyenv python binaries, and virtual environments in a sort of 'hierarchy' from global to local. When you install python and use pyenv, you're doing something similar to what I described over here with virtual environments: https://github.com/wehr-lab/autopilot/discussions/73#discussioncomment-358301 - You're adding a "shim" to the front of $PATH that tells the shell to look for the pyenv Python before other Python binaries. 

When you create a virtual environment, unless otherwise specified, it uses the python interpreter that was invoked to create it. When you activate a virtual environment, it puts its python interpreter at the front of `$PATH` as well, so if you had used python 3.8 to create a virtual environment, and then activate that environment, even though your pyenv version is set as 3.7.7 the shell will find the virtual environment python first. So what you would want to do here is create a new environment and use the `-p` flag to virtualenv, like this `python -m virtualenv -p python3.7 ./venv37` You can see more about how virtualenv handles finding Python interpreters in their documentation: https://virtualenv.pypa.io/en/latest/user_guide.html#python-discovery

There is also a pyenv plugin that makes virtualenvs a little easier to use here: https://github.com/pyenv/pyenv-virtualenv

Alternatively, it looks like you make a fresh clone of the repository, so if you wanted to use your python 3.8 environment, you would need to comment out the version requirement in the line i mention above again https://github.com/wehr-lab/autopilot/discussions/72#discussioncomment-358595

**Reply by neurodavidus** [reply]
OK,  I thought I made it but nope
I managed to install this pyenv plugin and create the correct environment with python3.7. (run into a couple of errors as it required modifying .bashrc). I was proud ...
Installed autopilot from source worked fine: no more complain about python 3.8
Autoconf ok but when I try to launch I got this error message. 
![image](https://user-images.githubusercontent.com/4154936/108541646-fc89d980-72e2-11eb-83d7-b107523a45d8.png)
this is weird because it looks that this file should have been created during the autoconfig ....

Any suggestion? thanks

**Reply by sneakers-the-rat** [reply]
that's a bug! looks like the place where the file was created was removed from the setup function when i rewrote it. Don't know how i could have missed that because i've made a bunch of new terminals since then... 

Anyway, it's been patched, should work now!! (though when creating the terminal for the first time it will have a blank screen which is super unfortunate, you just need to use "create pilot" from the file menu :)

Thanks again for letting me know!!!

PR: https://github.com/wehr-lab/autopilot/pull/79
PyPI: https://pypi.org/project/auto-pi-lot/0.3.5/
Release: https://github.com/wehr-lab/autopilot/releases/tag/v0.3.5

**Reply by neurodavidus** [reply]
I am glad I can help discover bugs while stumbling here and there. Autopilot installed and configured!!! Thanks
Next step running this blinking task: new discussions to come very soon :)
